docs/design/kmp-nversions.md

```diff
@@ -0,0 +1,106 @@
+
+## Problem Statement
+
+In version 1.3.0 and earlier, Ratify does not support the nVersionCount parameter for Key Management Provider (KMP) resources. This means that when a certificate or key is rotated, Ratify updates the cache with the new version and removes the previous one, which may not suit all use cases.
+
+For instance, if a user needs to retain the last three versions of a certificate or key in the cache, Ratify cannot meet this requirement without manually adjusting the KMP resource for each new version.
+
+By supporting nVersionCount, Ratify would allow users to specify how many versions of a certificate or key should be kept in the cache, eliminating the need for manual updates to the KMP resource.
+
+## Proposed Solution
+
+To address this challenge, this proposal suggests adding support for the `maxVersionCount` parameter to the KMP resource in Ratify. This parameter will allow users to specify the number of versions of a certificate or key that should be retained in the cache.
+
+When a new version of a certificate or key is created, Ratify will check the `maxVersionCount` parameter to determine how many versions should be retained in the cache. If the number of versions exceeds the specified count, Ratify will remove the oldest version from the cache.
+
+If a version is disabled, Ratify will remove it from the cache. This ensures that disabled versions are not retained in the cache, reducing the risk of using compromised keys or certificates being passed to the verifiers.
+
+Example: AKV KMP resource with `maxVersionCount` parameter
+
+```yaml
+apiVersion: config.ratify.deislabs.io/v1beta1
+kind: KeyManagementProvider
+metadata:
+  name: keymanagementprovider-akv
+spec:
+  type: azurekeyvault
+  refreshInterval: 1m
+  parameters:
+    vaultURI: https://yourkeyvault.vault.azure.net/
+    certificates:
+      - name: yourCertName
+        maxVersionCount: 3
+    tenantID:
+    clientID:
+```
+
+Example: AKV KMP resource status with multiple versions retained in the cache
+
+```yaml
+Status:
+  Issuccess:        true
+  Lastfetchedtime:  2024-10-02T14:58:54Z
+  Properties:
+    Certificates:
+      Last Refreshed:  2024-10-02T14:58:54Z
+      Name:            yourCertName
+      Version:         a1b2c3d4e5f67890abcdef1234567890
+      Status:          Enabled
+      Last Refreshed:  2024-10-02T14:58:54Z
+      Name:            yourCertName
+      Version:         0ff373a9259c4578a247cfd7861a8805
+      Status:          Disabled
+```
+
+## Implementation Details
+
+- Modify the KMP data structure to include the status of the version.
+    ```go
+    type KMPMapKey struct {
+    Name    string
+    Version string
+    Status  string // Enabled or Disabled
+    Created time.Time // Time the version was created used for determining the oldest version
+    }
+    ```
+- Add the `maxVersionCount` parameter to the KMP resource in Ratify.
+  - ensure the value cannot be less than 1 or a negative number
+  - default to 0 if not specified
```

Comment:
The parameter name `maxVersionCount` seems not intuitive. For example, the default value `0` means max version count is `0` , however, we will cache one version, the latest one, right? If the value is `1`, we cached two versions. Maybe using `maxPreviousVersionCount` or `previousVersionCount` is better? Additonally, using `max` is also a bit confusing for me, for example, if the value is `2`, does that mean at maximum, two versions are cached, but caching one version is still OK?


Since it means the cached versions, how about using the parameter name `cachedVersionCount` or `cachedPreviousVersionCount`?

Reply:
discussed during community meeting. we want to avoid 'cache' as this is implementation detail. We are proposing ,  "syncVersionCount", and given @duffney 's research that 3 is a common default. We propose to stay with 3 as the default.

Reply:
not a strong opinion, just posting a var name recommended by Chatgpt, `versionsFetchLimit`

Reply:
secret store uses `objectVersionHistory`, so perhaps a good name for us would be `certificateVersionHistory`? or simply `versionHistory`?

Reply:
discussed during community meeting, we would go with "versionHistory"

Reply:
Vote for `versionHistory`

Reply:
+1 for `versionHistory`